src/util/mod.rs

```diff
@@ -54,3 +59,25 @@ pub fn make_spirv<'a>(data: &'a [u8]) -> super::ShaderSource<'a> {
     );
     super::ShaderSource::SpirV(words)
 }
+
+/// Asynchronously read the contents of buffer.
+pub fn read_buffer(device: &super::Device, queue: &super::Queue, buffer: &super::Buffer, offset: super::BufferAddress, size: super::BufferAddress) -> impl Future<Output=Result<Vec<u8>, super::BufferAsyncError>> + Send {
```

Comment:
should accept a `BufferSlice`, makes things simpler

Reply:
Is there a way to go from a BufferSlice to the size that's simpler than this?
```
slice.size.map(Into::into).unwrap_or(slice.buffer.map_context.lock().total_size)
```

Reply:
We could place `total_size` on the buffer itself instead of the `map_context` to simplify this.
Also this code needs to take into account the offset:
```rust
match slice.size {
  Some(size) => size.into(),
  None => slice.buffer.map_context.lock().total_size - slice.offset,
}
```